calculator_test.go

```diff
@@ -2,23 +2,202 @@ package calculator_test
 
 import (
 	"calculator"
+	"math/rand"
 	"testing"
+	"time"
 )
 
+// testCase represents a case for testing non-error producing Calculator functions.
+type testCase struct {
+	a, b, want float64
+	name       string
+}
+
+// errorTestCase represents a case for testing Calculator functions that return errors.
+type errorTestCase struct {
+	a, b, want  float64
+	name        string
+	errExpected bool
+}
+
+// variadicTestCase represents a test case for a Calculator function that
+// accepts a variable number of inputs.
+type variadicTestCase struct {
+	inputs []float64
+	want   float64
+	name   string
+}
+
+// variadicErrorTestCase represents a test case for a Calculator function
+// that can accept a variable number of inputs and can return errors.
+type variadicErrorTestCase struct {
+	inputs      []float64
+	want        float64
+	name        string
+	errExpected bool
+}
+
+// evaluationTestCase represents a test case for any function which expects
+// a mathematical expression given as a string to be evaluated
+// and can return errors.
+type evaluationTestCase struct {
+	expression, name string
+	want             float64
+	errExpected      bool
+}
+
 func TestAdd(t *testing.T) {
 	t.Parallel()
-	var want float64 = 4
-	got := calculator.Add(2, 2)
-	if want != got {
-		t.Errorf("want %f, got %f", want, got)
+	testCases := []*variadicTestCase{
```

Comment:
Do we need _pointers_ to the test case structs? If not, we could just use a slice of values, couldn't we?

Reply:
You are exactly right, unless we need to modify any fields in the struct, there's really no point of passing a pointer to it!